### PR TITLE
Provide modules with information how much a quantity changed

### LINKED ISSUE #25981: ps_emailalerts - out-of-stock email is sent everytime the product is saved
#### Describe the bug

With combination products, if one of the combination is out-of-stock and I edit the product, everytime I save it, I'll receive an email.

#### Expected behavior

I should receive an out-of-stock email alert only once and not everytime I make a change on the product.

#### Steps to Reproduce

1. Install ps_emailalerts and Configure > Set threshold to 3
2. Edit a product with combination 
3. Set one combination quantity to 10, second combination to 1 > Save. You should receive email about low stock -> **OK**
3. Edit your product description > Save
4. You will receive again the out-of-stock alert -> **NOK**

**Screenshots**

Out-of-stock email sent for each combination :

Screenshot 2021-09-24 at 11 55 31

#### Additional information

* PrestaShop version: develop
* PHP version: 7.3

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR allows modules to know how much the quantity has changed during update. Not only the new number.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Partially fixes #25981
| How to test?      | Test using ps_qualityassurance and see that correct increments/decrements are provided in delta_quantity
| Possible impacts? | -

I named the parameter the same as in other parts of PS: https://github.com/PrestaShop/PrestaShop/blob/ad181f7fb89dea67dfebba779a6440854baa6d5b/src/Core/Stock/StockManager.php#L193

### CODE REVIEW

**Review by matks — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

I have a case non treated following these steps:
1. Install ps_emailalert and set the threshold to 3
2. configure the Email 
3. Go to BO > Catalog > Create new product and enable it (by default, quantity = 0)
4. There is no email out of stock 

Without PR: the email is successfully sent after creating a new product out of stock.
image

Ping @PrestaShop/product-team we need your confirmation about this behavior

Thanks!

**Comment by Hlavtox** [comment]
@khouloudbelguith Made a small mod, now it should send the email. But only if you have default activation state set to visible, because we don't send anything for disabled products.

**Comment by marionf** [comment]
@khouloudbelguith 

You are right, [the use case](https://github.com/PrestaShop/PrestaShop/pull/25984#issuecomment-931206590) you described should be fixed

**Comment by Hlavtox** [comment]
@marionf @khouloudbelguith The use case you mentioned should be fixed then.

File: classes/stock/StockAvailable.php [hunk 1]
```diff
@@ -598,6 +598,7 @@ public static function setQuantity($id_product, $id_product_attribute, $quantity
                     $stockManager->saveMovement($id_product, $id_product_attribute, $deltaQuantity);
                 }
             } else {
+                $deltaQuantity = null;
```
Comment (Progi1984) [inline classes/stock/StockAvailable.php]:
```suggestion
```

File: classes/stock/StockAvailable.php [hunk 2]
```diff
@@ -630,6 +631,7 @@ public static function setQuantity($id_product, $id_product_attribute, $quantity
                     'id_product' => $id_product,
                     'id_product_attribute' => $id_product_attribute,
                     'quantity' => $stock_available->quantity,
+                    'delta_quantity' => $deltaQuantity,
```
Comment (Progi1984) [inline classes/stock/StockAvailable.php]:
```suggestion
                    'delta_quantity' => $deltaQuantity ?? null,
```

**Review by Progi1984 — COMMENTED** [review]
@Hlavtox Thank you for your contribution. I have brought you a small comment.

**Review by kpodemski — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

Sorry, but I have still the same issue.
> Made a small mod, now it should send the email. But only if you have default activation state set to visible, because we don't send anything for disabled products.

Steps to reproduce the issue:

1. Install ps_emailalert and set the threshold to 3
2. configure the Email
3. Go to BO > Catalog > Create new product and enable it (by default, quantity = 0)
4. There is no email out of stock

I attached a screen record
https://www.awesomescreenshot.com/video/5794696?key=44bc78a7a24093a1bca30800e5bb5cf2

Other steps to reproduce the issue

1. Go to BO > Shop Parameters > Product Settings > Enable `Default activation status ` and save 
image
2. Go to BO > Product > Create new product and save
3. No email sent
4. If you try to generate combinations => no email sent also 
5. If you try to duplicate a product and enable it => no email sent

I attached a screen record
https://www.awesomescreenshot.com/video/5794578?key=9e005735b10030763049023b63d373d5

Thanks to check and feedback.

**Comment by Hlavtox** [comment]
@khouloudbelguith Did you use the current version of the related PR?

**Comment by Hlavtox** [comment]
@khouloudbelguith @Progi1984 I think we should discuss sending or not sending the emails in ps_emailalerts PR.
This PR provides modules with data and is working fine, so I think it should be merged...

**Comment by khouloudbelguith** [comment]
> @khouloudbelguith Did you use the current version of the related PR?

Hi @Hlavtox,

Yes, I'm using this PR: https://github.com/PrestaShop/ps_emailalerts/pull/91 while testing this PR: #25984

Thanks!

**Comment by Hlavtox** [comment]
@khouloudbelguith OK, I will deal with this module later. This PR is providing modules with data about quantity changes and I think it works as it should.

**Review by khouloudbelguith — DISMISSED** [review]
Hi @Hlavtox,

In the `How to test` part, `Install both this and related PR`.
So I installed this PR using the PR module 
https://github.com/PrestaShop/ps_emailalerts/pull/91

I still have the same issue:
1. Go to BO > Catalog > Products
2. Create a product
3. Enable it
4. See error => there is no mail out of stock sent

I attached a screen

https://user-images.githubusercontent.com/16067358/152825536-19e4f2b7-2611-4107-b9d7-79a02421285c.mp4

**Without PR**

https://user-images.githubusercontent.com/16067358/152826153-320265f2-35b2-4979-a778-e0a9482aa758.mp4


As confirmed by the Product team, this issue should be fixed:
https://github.com/PrestaShop/PrestaShop/pull/25984#issuecomment-941004293

Thanks!

**Comment by kpodemski** [comment]
Hello @khouloudbelguith @Hlavtox

I really don't think it requires more checks than having:

`PrestaShopLogger::addLog('Delta from '.date('d.m.Y H:i:s').': '.$params['delta_quantity']);`

inside `ps_qualityassurance` for the hook: `actionUpdateQuantity`.

This PR is about adding this information for the developer, and it's what it does. We can test PR related to `ps_emailalerts` later on.

I suggest the following:
- make sure that the `delta_quantity` information is correct using `ps_qualityassurance` module @khouloudbelguith 
- update the description and issue number to not involve `ps_emailalerts` module for now @Hlavtox 

What do you think?

**Comment by Hlavtox** [comment]
@kpodemski I agree of course, thank you :-)

**Comment by Progi1984** [comment]
@Hlavtox Could you update the `How to test?` for @PrestaShop/qa-functional team?

**Review by khouloudbelguith — COMMENTED** [review]
H @Hlavtox, @kpodemski,

> I really don't think it requires more checks than having:
> 
> `PrestaShopLogger::addLog('Delta from '.date('d.m.Y H:i:s').': '.$params['delta_quantity']);`
> 
> inside `ps_qualityassurance` for the hook: `actionUpdateQuantity`.
> 
> This PR is about adding this information for the developer, and it's what it does. We can test PR related to `ps_emailalerts` later on.
> 
> I suggest the following:
> 
> * make sure that the `delta_quantity` information is correct using `ps_qualityassurance` module @khouloudbelguith
> * update the description and issue number to not involve `ps_emailalerts` module for now @Hlavtox
> 
> What do you think?

Thanks for your feedbacks.
I followed these steps:

1. Install this PR and Install the ps_qualityassurance module
2. Add the hook `actionUpdateQuantity` 
image
3. Create a new product or edit a product
4. In the BO > Advanced Parameters > Logs => The delta quantity is well displayed
image

But as discussed with @Robin-Fischer-PS, thanks to update the `How to test `of this PR because the other [PR](https://github.com/PrestaShop/ps_emailalerts/pull/91) is NOK as described [here](https://github.com/PrestaShop/PrestaShop/pull/25984#pullrequestreview-874914197).
Also, thanks to provide us with more details about the possible impacts of this PR, so we can test and check all possible scenarios.

Thank you!

**Comment by Hlavtox** [comment]
@khouloudbelguith Hello! PR description has been updated, we can finish that PR later, when this is merged. :-)

**Review by khouloudbelguith — COMMENTED** [review]
Hi @Hlavtox,

Thanks for your update.
I found this issue

1. Install the ps_qualityassurance & add the hook `actionUpdateQuantity` ( `PrestaShopLogger::addLog('Delta from '.date('d.m.Y H:i:s').': '.$params['delta_quantity']);`)
2. Go to BO > Advanced Parameters > New & Experimental Features page && enable `New product page - Single store`
3. Go to Catalog > Products > Edit a product 
4. Change a quantity & save
5. Go to BO > Advanced Parmeters > Logs page 
6. check that no log added `Delta from 20.06.2022 16:46:11: 2`

https://user-images.githubusercontent.com/16067358/174630217-e712ee8a-1866-4fda-85fe-f531c8291a27.mp4

I tried with Product V1 => OK :heavy_check_mark: 
I tried to update from the Catalog > Stocks page => ok
I tried to update the quantity from the BO > Order view page => ok
I tried to refund an order => ok
I tried to create an order => ok

Thanks!

**Comment by matks** [comment]
@Hlavtox friendly reminder 😄 QA team detected issues on this contribution

**Comment by Hlavtox** [comment]
Hi @khouloudbelguith and @PrestaShop/prestashop-maintainers 

- I rebased this PR.
- Added a new hook call also to the new product page. It follows the same logic and is called on the same space as the old hook, just before `StockManager::saveMovement();`.
- I added id_shop to another hook call, to unify them.

File: src/Adapter/Product/Stock/Update/ProductStockUpdater.php [hunk 3]
```diff
@@ -309,5 +310,16 @@ private function saveMovement(StockAvailable $stockAvailable, StockModification
                 'id_shop' => (int) $stockAvailable->id_shop,
             ]
         );
+
+        Hook::exec(
```
Comment (kpodemski) [inline src/Adapter/Product/Stock/Update/ProductStockUpdater.php]: Wow, that's quite a big issue if there was no hook previously

**Review by kpodemski — DISMISSED** [review]

**Review by marsaldev — DISMISSED** [review]

File: src/Adapter/Product/Stock/Update/ProductStockUpdater.php (hunk 3 above)
Comment (FabienPapet) [reply src/Adapter/Product/Stock/Update/ProductStockUpdater.php]: There should be dependency injection here.

**Review by FabienPapet — CHANGES_REQUESTED** [review]

**Review by kpodemski — COMMENTED** [review]
@Hlavtox 

@FabienPapet is right. I didn't notice it's a Symfony service, you can safely add the next argument in the constructor and inject HookDispatcher 👍

**Comment by Hlavtox** [comment]
Hi @kpodemski @FabienPapet @marsaldev, I changed the implementation to DI as requested 👍
So if we can please reapprove and re-assign Waiting for QA.

**Review by FabienPapet — APPROVED** [review]

File: src/Adapter/Product/Stock/Update/ProductStockUpdater.php [hunk 4]
```diff
@@ -309,5 +316,16 @@ private function saveMovement(StockAvailable $stockAvailable, StockModification
                 'id_shop' => (int) $stockAvailable->id_shop,
             ]
         );
+
+        $this->hookManager->exec(
```
Comment (jolelievre) [inline src/Adapter/Product/Stock/Update/ProductStockUpdater.php:320]: Nice catch!! I didn't notice there was a missing hook from the static method, thanks

**Review by jolelievre — APPROVED** [review]
Thanks @Hlavtox

**Review by kpodemski — APPROVED** [review]

**Review by hibatallahAouadni — APPROVED** [review]
Hello @Hlavtox 

LGTM, QA :heavy_check_mark: 
image

Edit product qty from:

- Product V1 :heavy_check_mark: 
- Product V2 :heavy_check_mark: 
- Stock management :heavy_check_mark: 
- Create an Order (FO) :heavy_check_mark: 
- Create an Order (BO) :heavy_check_mark: 

Thanks!

**Comment by kpodemski** [comment]
Thank you @Hlavtox @hibatallahAouadni :)

**Pull request merged**